File: rs/tests/metadata_test.rs
/// Tests for metadata/schema, security limits, list_children, read_entry_prefix,
/// and path helpers added to shard_v2.
use shard_format::{
    EntryMeta, FLAG_HAS_SCHEMA, HEADER_SIZE, INDEX_ENTRY_SIZE, MAX_ENTRY_COUNT, ROLE_MOSH,
    ShardMetadata, ShardV2Reader, ShardV2Writer, join_path, path_base, path_parent, split_path,
};

// ============================================================
// Helper
// ============================================================

fn make_shard(entries: &[(&str, &[u8])]) -> Vec<u8> {
    let mut w = ShardV2Writer::new(ROLE_MOSH);
    for (name, data) in entries {
        w.write_entry(name, data);
    }
    w.to_bytes()
}

fn make_shard_with_meta(entries: &[(&str, &[u8])], meta: ShardMetadata) -> Vec<u8> {
    let mut w = ShardV2Writer::new(ROLE_MOSH);
    for (name, data) in entries {
        w.write_entry(name, data);
    }
    w.set_metadata(meta);
    w.to_bytes()
}

// ============================================================
// Path helpers
// ============================================================

#[test]
fn split_path_simple() {
    assert_eq!(split_path("layer.0/weight"), vec!["layer.0", "weight"]);
}

#[test]
fn split_path_deep() {
    assert_eq!(split_path("a/b/c/d"), vec!["a", "b", "c", "d"]);
}

#[test]
fn split_path_no_slash() {
    assert_eq!(split_path("embed"), vec!["embed"]);
}

#[test]
fn split_path_empty() {
    let result: Vec<&str> = split_path("");
    assert!(result.is_empty());
}

#[test]
fn join_path_two_parts() {
    assert_eq!(join_path(&["layer.0", "weight"]), "layer.0/weight");
}

#[test]
fn join_path_single() {
    assert_eq!(join_path(&["embed"]), "embed");
}

#[test]
fn join_path_multi() {
    assert_eq!(join_path(&["a", "b", "c"]), "a/b/c");
}

#[test]
fn path_parent_with_slash() {
    assert_eq!(path_parent("layer.0/weight"), "layer.0");
}

#[test]
fn path_parent_deep() {
    assert_eq!(path_parent("a/b/c"), "a/b");
}

#[test]
fn path_parent_no_slash() {
    assert_eq!(path_parent("embed"), "");
}

#[test]
fn path_base_with_slash() {
    assert_eq!(path_base("layer.0/weight"), "weight");
}

#[test]
fn path_base_deep() {
    assert_eq!(path_base("a/b/c"), "c");
}

#[test]
fn path_base_no_slash() {
    assert_eq!(path_base("embed"), "embed");
}

// ============================================================
// list_children
// ============================================================

fn make_layered_shard() -> Vec<u8> {
    make_shard(&[
        ("layer.0/weight", b"w0"),
        ("layer.0/bias", b"b0"),
        ("layer.1/weight", b"w1"),
        ("embed", b"tok"),
    ])
}

#[test]
fn list_children_exact_prefix_with_slash() {
    let buf = make_layered_shard();
    let r = ShardV2Reader::from_bytes(buf).unwrap();
    let mut result = r.list_children("layer.0/");
    result.sort();
    assert_eq!(result, vec!["layer.0/bias", "layer.0/weight"]);
}

#[test]
fn list_children_empty_prefix_returns_top_level() {
    let buf = make_layered_shard();
    let r = ShardV2Reader::from_bytes(buf).unwrap();
    let result = r.list_children("");
    assert!(result.contains(&"layer.0/".to_string()));
    assert!(result.contains(&"layer.1/".to_string()));
    assert!(result.contains(&"embed".to_string()));
    // No duplicates
    let mut sorted = result.clone();
    sorted.dedup();
    assert_eq!(result.len(), sorted.len());
}

#[test]
fn list_children_partial_prefix() {
    let buf = make_layered_shard();
    let r = ShardV2Reader::from_bytes(buf).unwrap();
    let mut result = r.list_children("layer.");
    result.sort();
    assert_eq!(result, vec!["layer.0/", "layer.1/"]);
}

#[test]
fn list_children_nonexistent_prefix() {
    let buf = make_shard(&[("a/b", b"x")]);
    let r = ShardV2Reader::from_bytes(buf).unwrap();
    let result = r.list_children("nonexistent/");
    assert!(result.is_empty());
}

#[test]
fn list_children_dedup_directories() {
    let buf = make_shard(&[
        ("a/b", b"1"),
        ("a/c", b"2"),
        ("a/d", b"3"),
    ]);
    let r = ShardV2Reader::from_bytes(buf).unwrap();
    let result = r.list_children("");
    // "a/" should appear exactly once
    assert_eq!(result.iter().filter(|x| x.as_str() == "a/").count(), 1);
}

#[test]
fn list_children_hierarchical_three_levels() {
    let buf = make_shard(&[
        ("a/b/c", b"1"),
        ("a/b/d", b"2"),
        ("a/e", b"3"),
        ("f", b"4"),
    ]);
    let r = ShardV2Reader::from_bytes(buf).unwrap();

    let top = r.list_children("");
    assert!(top.contains(&"a/".to_string()));
    assert!(top.contains(&"f".to_string()));

    let under_a = r.list_children("a/");
    assert!(under_a.contains(&"a/b/".to_string()));
    assert!(under_a.contains(&"a/e".to_string()));

    let mut under_ab = r.list_children("a/b/");
    under_ab.sort();
    assert_eq!(under_ab, vec!["a/b/c", "a/b/d"]);
}

// ============================================================
// read_entry_prefix
// ============================================================

const PAYLOAD: &[u8] = b"Hello World, this is a test payload!";

#[test]
fn read_entry_prefix_first_n_bytes() {
    let buf = make_shard(&[("entry", PAYLOAD)]);
    let r = ShardV2Reader::from_bytes(buf).unwrap();
    let got = r.read_entry_prefix(0, 5).unwrap();
    assert_eq!(got, b"Hello");
}

#[test]
fn read_entry_prefix_full_entry() {
    let buf = make_shard(&[("entry", PAYLOAD)]);
    let r = ShardV2Reader::from_bytes(buf).unwrap();
    let got = r.read_entry_prefix(0, PAYLOAD.len()).unwrap();
    assert_eq!(got, PAYLOAD);
}

#[test]
fn read_entry_prefix_exceeds_entry_length() {
    let buf = make_shard(&[("entry", PAYLOAD)]);
    let r = ShardV2Reader::from_bytes(buf).unwrap();
    let got = r.read_entry_prefix(0, 100_000).unwrap();
    assert_eq!(got, PAYLOAD);
}

#[test]
fn read_entry_prefix_zero_bytes() {
    let buf = make_shard(&[("entry", PAYLOAD)]);
    let r = ShardV2Reader::from_bytes(buf).unwrap();
    let got = r.read_entry_prefix(0, 0).unwrap();
    assert!(got.is_empty());
}

#[test]
fn read_entry_prefix_one_byte() {
    let buf = make_shard(&[("entry", PAYLOAD)]);
    let r = ShardV2Reader::from_bytes(buf).unwrap();
    let got = r.read_entry_prefix(0, 1).unwrap();
    assert_eq!(got, b"H");
}

// ============================================================
// ShardMetadata roundtrip
// ============================================================

#[test]
fn metadata_basic_roundtrip() {
    let mut meta = ShardMetadata::default();
    meta.schema_version = "shard-v2.1".to_string();
    meta.producer = "test-producer".to_string();
    meta.description = "A test shard".to_string();
    meta.created_at = "2026-02-21T00:00:00Z".to_string();
    meta.tags = vec!["ml".to_string(), "test".to_string()];

    let buf = make_shard_with_meta(&[("data", b"hello")], meta);
    let r = ShardV2Reader::from_bytes(buf).unwrap();

    // FLAG_HAS_SCHEMA must be set
    assert!(r.header().flags & FLAG_HAS_SCHEMA != 0);
    assert!(r.header().schema_offset > 0);

    let got = r.read_metadata().unwrap().expect("metadata should be Some");
    assert_eq!(got.producer, "test-producer");
    assert_eq!(got.description, "A test shard");
    assert_eq!(got.created_at, "2026-02-21T00:00:00Z");
    assert_eq!(got.tags, vec!["ml", "test"]);
    assert_eq!(got.schema_version, "shard-v2.1");
}

#[test]
fn metadata_entry_metadata_roundtrip() {
    let mut em = EntryMeta::default();
    em.content_type = "tensor/float32".to_string();
    em.tags = vec!["weight".to_string()];
    em.description = "Model weights".to_string();
    em.extra.insert(
        "shape".to_string(),
        serde_json::json!([256, 256]),
    );

    let mut meta = ShardMetadata::default();
    meta.producer = "pytorch-exporter".to_string();
    meta.entry_metadata.insert("model/weight".to_string(), em);

    let buf = make_shard_with_meta(&[("model/weight", &[0u8; 16])], meta);
    let r = ShardV2Reader::from_bytes(buf).unwrap();
    let got = r.read_metadata().unwrap().expect("metadata should be Some");

    assert!(got.entry_metadata.contains_key("model/weight"));
    let got_em = &got.entry_metadata["model/weight"];
    assert_eq!(got_em.content_type, "tensor/float32");
    assert_eq!(got_em.tags, vec!["weight"]);
    assert_eq!(got_em.description, "Model weights");
    assert_eq!(got_em.extra["shape"], serde_json::json!([256, 256]));
}

#[test]
fn metadata_no_metadata_returns_none() {
    let buf = make_shard(&[("data", b"hello")]);
    let r = ShardV2Reader::from_bytes(buf).unwrap();
    assert_eq!(r.header().schema_offset, 0);
    assert!(r.read_metadata().unwrap().is_none());
}

#[test]
fn metadata_extra_fields() {
    let mut meta = ShardMetadata::default();
    meta.extra.insert("model_type".to_string(), serde_json::json!("transformer"));
    meta.extra.insert("num_layers".to_string(), serde_json::json!(12));

    let buf = make_shard_with_meta(&[("cfg", b"x")], meta);
    let r = ShardV2Reader::from_bytes(buf).unwrap();
    let got = r.read_metadata().unwrap().expect("metadata should be Some");
    assert_eq!(got.extra["model_type"], serde_json::json!("transformer"));
    assert_eq!(got.extra["num_layers"], serde_json::json!(12));
}

#[test]
fn metadata_source_uri_schema_uri() {
    let mut meta = ShardMetadata::default();
    meta.source_uri = "s3://bucket/model.safetensors".to_string();
    meta.schema_uri = "https://example.com/schema.json".to_string();

    let buf = make_shard_with_meta(&[("x", b"y")], meta);
    let r = ShardV2Reader::from_bytes(buf).unwrap();
    let got = r.read_metadata().unwrap().expect("metadata should be Some");
    assert_eq!(got.source_uri, "s3://bucket/model.safetensors");
    assert_eq!(got.schema_uri, "https://example.com/schema.json");
}

#[test]
fn metadata_default_schema_version() {
    let meta = ShardMetadata::default();
    let buf = make_shard_with_meta(&[("x", b"y")], meta);
    let r = ShardV2Reader::from_bytes(buf).unwrap();
    let got = r.read_metadata().unwrap().expect("metadata should be Some");
    assert_eq!(got.schema_version, "shard-v2.1");
}

#[test]
fn security_schema_offset_beyond_file_is_rejected() {
    let mut buf = make_shard_with_meta(&[("data", b"hello")], ShardMetadata::default());
    let invalid = (buf.len() as u64) + 1;
    buf[32..40].copy_from_slice(&invalid.to_le_bytes());

    let err = ShardV2Reader::from_bytes(buf).unwrap_err();
    let msg = format!("{}", err);
    assert!(msg.contains("schema_offset"), "expected schema_offset error, got: {}", msg);
}

#[test]
fn security_schema_offset_overlapping_data_is_rejected() {
    let mut buf = make_shard_with_meta(&[("data", b"hello world")], ShardMetadata::default());
    let entry_data_offset_off = HEADER_SIZE + 16;
    let data_offset = u64::from_le_bytes(
        buf[entry_data_offset_off..entry_data_offset_off + 8]
            .try_into()
            .unwrap(),
    );
    buf[32..40].copy_from_slice(&data_offset.to_le_bytes());

    let err = ShardV2Reader::from_bytes(buf).unwrap_err();
    let msg = format!("{}", err);
    assert!(msg.contains("schema_offset"), "expected schema_offset error, got: {}", msg);
}

// ============================================================
// Security limits
// ============================================================

fn make_corrupt_header(entry_count: u32, total_file_size_override: Option<u64>) -> Vec<u8> {
    let mut buf = vec![0u8; HEADER_SIZE];
    buf[0..4].copy_from_slice(b"SHRD");
    buf[4] = 0x02; // version
    buf[5] = 1;    // role
    buf[6..8].copy_from_slice(&0x00A2u16.to_le_bytes()); // flags
    buf[8] = 64;   // alignment
    buf[9] = 0;    // compression
    buf[10..12].copy_from_slice(&(INDEX_ENTRY_SIZE as u16).to_le_bytes());
    buf[12..16].copy_from_slice(&entry_count.to_le_bytes());

    let st_off = HEADER_SIZE as u64;
    let ds_off = st_off;
    buf[16..24].copy_from_slice(&st_off.to_le_bytes()); // string_table_offset
    buf[24..32].copy_from_slice(&ds_off.to_le_bytes()); // data_section_offset
    buf[32..40].copy_from_slice(&0u64.to_le_bytes());   // schema_offset = 0

    let tfs = total_file_size_override.unwrap_or(HEADER_SIZE as u64);
    buf[40..48].copy_from_slice(&tfs.to_le_bytes());    // total_file_size

    buf
}

#[test]
fn security_entry_count_too_large() {
    let too_many = (MAX_ENTRY_COUNT + 1) as u32;
    let buf = make_corrupt_header(too_many, Some(HEADER_SIZE as u64));
    let err = ShardV2Reader::from_bytes(buf).unwrap_err();
    let msg = format!("{}", err);
    assert!(
        msg.contains("MAX_ENTRY_COUNT"),
        "expected MAX_ENTRY_COUNT in error, got: {}",
        msg
    );
}

#[test]
fn security_total_file_size_mismatch() {
    // Build a valid shard then corrupt total_file_size
    let mut buf = make_shard(&[("x", b"hello")]);
    // Overwrite total_file_size at offset 40
    let corrupt_size = 99999u64;
    buf[40..48].copy_from_slice(&corrupt_size.to_le_bytes());
    let err = ShardV2Reader::from_bytes(buf).unwrap_err();
    let msg = format!("{}", err);
    assert!(
        msg.contains("total_file_size"),
        "expected total_file_size in error, got: {}",
        msg
    );
}

#[test]
fn security_data_offset_before_data_section() {
    // Build a valid shard, then set entry 0 data_offset to 0 (before header)
    let mut buf = make_shard(&[("x", b"hello")]);
    let entry_off = HEADER_SIZE + 16; // data_offset field is at index+16
    buf[entry_off..entry_off + 8].copy_from_slice(&0u64.to_le_bytes());
    // Keep total_file_size consistent
    let tfs = buf.len() as u64;
    buf[40..48].copy_from_slice(&tfs.to_le_bytes());
    assert!(ShardV2Reader::from_bytes(buf).is_err());
}

#[test]
fn security_data_extends_past_end_of_file() {
    // Build a valid shard, then set entry 0 data_offset to huge value
    let mut buf = make_shard(&[("x", b"hello")]);
    let entry_off = HEADER_SIZE + 16;
    buf[entry_off..entry_off + 8].copy_from_slice(&0xFFFF_FFFFu64.to_le_bytes());
    let tfs = buf.len() as u64;
    buf[40..48].copy_from_slice(&tfs.to_le_bytes());
    assert!(ShardV2Reader::from_bytes(buf).is_err());
}

File: ts/src/__tests__/metadata.test.ts
/**
 * Tests for metadata/schema, security limits, listChildren, readEntryPrefix,
 * and path helpers added to shard TypeScript.
 */

import { describe, it, expect, beforeAll } from 'vitest';

import {
  ShardReader,
  ShardWriter,
  ShardMetadata,
  EntryMeta,
  initXxhash,
  FLAG_HAS_SCHEMA,
  MAX_ENTRY_COUNT,
  MAX_INDEX_SIZE,
  MAX_STRING_TABLE_SIZE,
  splitPath,
  joinPath,
  pathParent,
  pathBase,
  serializeMetadata,
  deserializeMetadata,
} from '../index.js';

beforeAll(async () => {
  await initXxhash();
});

// ============================================================
// Helper
// ============================================================

function makeShardBuffer(
  entries: Array<[string, Buffer]>,
  metadata?: ShardMetadata,
): Buffer {
  const w = new ShardWriter();
  for (const [name, data] of entries) {
    w.writeEntry(name, data);
  }
  if (metadata !== undefined) {
    w.setMetadata(metadata);
  }
  return w.toBuffer();
}

// ============================================================
// Path helpers
// ============================================================

describe('path helpers', () => {
  it('splitPath simple', () => {
    expect(splitPath('layer.0/weight')).toEqual(['layer.0', 'weight']);
  });

  it('splitPath deep', () => {
    expect(splitPath('a/b/c/d')).toEqual(['a', 'b', 'c', 'd']);
  });

  it('splitPath no slash', () => {
    expect(splitPath('embed')).toEqual(['embed']);
  });

  it('splitPath empty', () => {
    expect(splitPath('')).toEqual([]);
  });

  it('joinPath two parts', () => {
    expect(joinPath('layer.0', 'weight')).toBe('layer.0/weight');
  });

  it('joinPath single', () => {
    expect(joinPath('embed')).toBe('embed');
  });

  it('joinPath multi', () => {
    expect(joinPath('a', 'b', 'c')).toBe('a/b/c');
  });

  it('pathParent with slash', () => {
    expect(pathParent('layer.0/weight')).toBe('layer.0');
  });

  it('pathParent deep', () => {
    expect(pathParent('a/b/c')).toBe('a/b');
  });

  it('pathParent no slash', () => {
    expect(pathParent('embed')).toBe('');
  });

  it('pathBase with slash', () => {
    expect(pathBase('layer.0/weight')).toBe('weight');
  });

  it('pathBase deep', () => {
    expect(pathBase('a/b/c')).toBe('c');
  });

  it('pathBase no slash', () => {
    expect(pathBase('embed')).toBe('embed');
  });
});

// ============================================================
// listChildren
// ============================================================

describe('listChildren', () => {
  it('exact prefix with trailing slash', () => {
    const buf = makeShardBuffer([
      ['layer.0/weight', Buffer.from('w0')],
      ['layer.0/bias', Buffer.from('b0')],
      ['layer.1/weight', Buffer.from('w1')],
      ['embed', Buffer.from('tok')],
    ]);
    const r = new ShardReader(buf);
    const result = r.listChildren('layer.0/');
    expect(result.sort()).toEqual(['bias', 'weight']);
  });

  it('empty prefix returns top-level children', () => {
    const buf = makeShardBuffer([
      ['layer.0/weight', Buffer.from('w0')],
      ['layer.0/bias', Buffer.from('b0')],
      ['layer.1/weight', Buffer.from('w1')],
      ['embed', Buffer.from('tok')],
    ]);
    const r = new ShardReader(buf);
    const result = r.listChildren('');
    expect(result).toContain('layer.0');
    expect(result).toContain('layer.1');
    expect(result).toContain('embed');
    // No duplicates
    expect(result.length).toBe(new Set(result).size);
  });

  it('partial prefix returns directory entries', () => {
    const buf = makeShardBuffer([
      ['layer.0/weight', Buffer.from('w0')],
      ['layer.0/bias', Buffer.from('b0')],
      ['layer.1/weight', Buffer.from('w1')],
      ['embed', Buffer.from('tok')],
    ]);
    const r = new ShardReader(buf);
    const result = r.listChildren('layer.');
    expect(result.sort()).toEqual(['0', '1']);
  });

  it('nonexistent prefix returns empty', () => {
    const buf = makeShardBuffer([
      ['a/b', Buffer.from('x')],
    ]);
    const r = new ShardReader(buf);
    expect(r.listChildren('nonexistent/')).toEqual([]);
  });

  it('deduplicated directories', () => {
    const buf = makeShardBuffer([
      ['a/b', Buffer.from('1')],
      ['a/c', Buffer.from('2')],
      ['a/d', Buffer.from('3')],
    ]);
    const r = new ShardReader(buf);
    // Under "" we should see "a" only once
    const result = r.listChildren('');
    expect(result.filter(x => x === 'a').length).toBe(1);
  });

  it('hierarchical three levels', () => {
    const buf = makeShardBuffer([
      ['a/b/c', Buffer.from('1')],
      ['a/b/d', Buffer.from('2')],
      ['a/e', Buffer.from('3')],
      ['f', Buffer.from('4')],
    ]);
    const r = new ShardReader(buf);

    const top = r.listChildren('');
    expect(top).toContain('a');
    expect(top).toContain('f');

    const underA = r.listChildren('a/');
    expect(underA).toContain('b');
    expect(underA).toContain('e');

    const underAB = r.listChildren('a/b/');
    expect(underAB.sort()).toEqual(['c', 'd']);
  });
});

// ============================================================
// readEntryPrefix
// ============================================================

describe('readEntryPrefix', () => {
  const data = Buffer.from('Hello World, this is a test payload!');

  it('returns first N bytes', () => {
    const buf = makeShardBuffer([['entry', data]]);
    const r = new ShardReader(buf);
    expect(r.readEntryPrefix(0, 5)).toEqual(Buffer.from('Hello'));
  });

  it('returns full entry when maxBytes equals entry length', () => {
    const buf = makeShardBuffer([['entry', data]]);
    const r = new ShardReader(buf);
    expect(r.readEntryPrefix(0, data.length)).toEqual(data);
  });

  it('returns full entry when maxBytes exceeds entry length', () => {
    const buf = makeShardBuffer([['entry', data]]);
    const r = new ShardReader(buf);
    expect(r.readEntryPrefix(0, 100_000)).toEqual(data);
  });

  it('returns empty buffer for maxBytes = 0', () => {
    const buf = makeShardBuffer([['entry', data]]);
    const r = new ShardReader(buf);
    expect(r.readEntryPrefix(0, 0)).toEqual(Buffer.alloc(0));
  });

  it('returns one byte', () => {
    const buf = makeShardBuffer([['entry', data]]);
    const r = new ShardReader(buf);
    expect(r.readEntryPrefix(0, 1)).toEqual(Buffer.from('H'));
  });

  it('throws on out-of-bounds index', () => {
    const buf = makeShardBuffer([['entry', data]]);
    const r = new ShardReader(buf);
    expect(() => r.readEntryPrefix(99, 10)).toThrow();
  });
});

// ============================================================
// ShardMetadata roundtrip
// ============================================================

describe('ShardMetadata roundtrip', () => {
  it('basic metadata survives write/read', () => {
    const meta: ShardMetadata = {
      schemaVersion: 'shard-v2.1',
      producer: 'test-producer',
      description: 'A test shard',
      createdAt: '2026-02-21T00:00:00Z',
      tags: ['ml', 'test'],
    };
    const buf = makeShardBuffer([['data', Buffer.from('hello')]], meta);
    const r = new ShardReader(buf);
    expect(r.header().flags & FLAG_HAS_SCHEMA).toBeTruthy();
    expect(Number(r.header().schemaOffset)).toBeGreaterThan(0);

    const got = r.readMetadata();
    expect(got).not.toBeNull();
    expect(got!.producer).toBe('test-producer');
    expect(got!.description).toBe('A test shard');
    expect(got!.createdAt).toBe('2026-02-21T00:00:00Z');
    expect(got!.tags).toEqual(['ml', 'test']);
    expect(got!.schemaVersion).toBe('shard-v2.1');
  });

  it('entry metadata roundtrip', () => {
    const em: EntryMeta = {
      contentType: 'tensor/float32',
      tags: ['weight'],
      description: 'Model weights',
      extra: { shape: [256, 256] },
      codec: 'cowrie-gen2',
      codecVersion: '2',
      schemaFingerprint: 'sha256:weights',
      semanticType: 'tensor',
      canonicalHash: 'sha256:canonical',
      baseHash: 'sha256:base',
      rowCount: 256,
      shape: [256, 256],
      stats: { min: -1, max: 1 },
    };
    const meta: ShardMetadata = {
      schemaVersion: 'shard-v2.1',
      producer: 'pytorch-exporter',
      entryMetadata: { 'model/weight': em },
    };
    const buf = makeShardBuffer([['model/weight', Buffer.alloc(16)]], meta);
    const r = new ShardReader(buf);
    const got = r.readMetadata();
    expect(got).not.toBeNull();
    expect(got!.entryMetadata).toBeDefined();
    const gotEm = got!.entryMetadata!['model/weight'];
    expect(gotEm).toBeDefined();
    expect(gotEm.contentType).toBe('tensor/float32');
    expect(gotEm.tags).toEqual(['weight']);
    expect(gotEm.description).toBe('Model weights');
    expect(gotEm.extra).toEqual({ shape: [256, 256] });
    expect(gotEm.codec).toBe('cowrie-gen2');
    expect(gotEm.codecVersion).toBe('2');
    expect(gotEm.schemaFingerprint).toBe('sha256:weights');
    expect(gotEm.semanticType).toBe('tensor');
    expect(gotEm.canonicalHash).toBe('sha256:canonical');
    expect(gotEm.baseHash).toBe('sha256:base');
    expect(gotEm.rowCount).toBe(256);
    expect(gotEm.shape).toEqual([256, 256]);
    expect(gotEm.stats).toEqual({ min: -1, max: 1 });
  });

  it('no metadata returns null', () => {
    const buf = makeShardBuffer([['data', Buffer.from('hello')]]);
    const r = new ShardReader(buf);
    expect(Number(r.header().schemaOffset)).toBe(0);
    expect(r.readMetadata()).toBeNull();
  });

  it('metadata with extra fields', () => {
    const meta: ShardMetadata = {
      schemaVersion: 'shard-v2.1',
      extra: { model_type: 'transformer', num_layers: 12 },
    };
    const buf = makeShardBuffer([['cfg', Buffer.from('x')]], meta);
    const r = new ShardReader(buf);
    const got = r.readMetadata();
    expect(got).not.toBeNull();
    expect(got!.extra!['model_type']).toBe('transformer');
    expect(got!.extra!['num_layers']).toBe(12);
  });

  it('metadata source_uri and schema_uri roundtrip', () => {
    const meta: ShardMetadata = {
      schemaVersion: 'shard-v2.1',
      sourceUri: 's3://bucket/model.safetensors',
      schemaUri: 'https://example.com/schema.json',
    };
    const buf = makeShardBuffer([['x', Buffer.from('y')]], meta);
    const r = new ShardReader(buf);
    const got = r.readMetadata();
    expect(got!.sourceUri).toBe('s3://bucket/model.safetensors');
    expect(got!.schemaUri).toBe('https://example.com/schema.json');
  });

  it('serializeMetadata / deserializeMetadata are inverses', () => {
    const meta: ShardMetadata = {
      schemaVersion: 'shard-v2.1',
      producer: 'unit-test',
      tags: ['a', 'b'],
      extra: { n: 42 },
      entryMetadata: {
        'foo/bar': { contentType: 'text/plain', description: 'test' },
      },
    };
    const serialized = serializeMetadata(meta);
    const got = deserializeMetadata(serialized);
    expect(got.producer).toBe('unit-test');
    expect(got.tags).toEqual(['a', 'b']);
    expect(got.extra!['n']).toBe(42);
    expect(got.entryMetadata!['foo/bar'].contentType).toBe('text/plain');
    expect(got.entryMetadata!['foo/bar'].description).toBe('test');
  });

  it('profile metadata roundtrip', () => {
    const meta: ShardMetadata = {
      schemaVersion: 'shard-v2.1',
      profile: 'sampleshard.v1',
      sampleShard: {
        datasetName: 'mnist-train',
        sampleIdType: 'uint64',
        keyEncoding: 'decimal-string',
        sampleCount: 60000,
        datasetSchema: { input: 'tensor[u8,28,28]', target: 'uint8' },
        splits: { train: { start: 0, end: 59999 } },
        labelMap: { '0': 'zero', '1': 'one' },
        featureStats: { input: { mean: 0.1307, std: 0.3081 } },
      },
      manifest: {
        files: [
          {
            uri: 's3://bucket/train-000.smpl',
            sha256: 'abc123',
            role: 'sample',
            profile: 'sampleshard.v1',
            startKey: '0',
            endKey: '59999',
            entryCount: 60000,
          },
        ],
        partitions: { train: ['train-000.smpl'] },
      },
    };

    const serialized = serializeMetadata(meta);
    const got = deserializeMetadata(serialized);
    expect(got.profile).toBe('sampleshard.v1');
    expect(got.sampleShard?.datasetName).toBe('mnist-train');
    expect(got.sampleShard?.keyEncoding).toBe('decimal-string');
    expect(got.sampleShard?.sampleCount).toBe(60000);
    expect(got.manifest?.files?.[0].uri).toBe('s3://bucket/train-000.smpl');
    expect(got.manifest?.partitions?.train).toEqual(['train-000.smpl']);
  });
});

// ============================================================
// Security limits
// ============================================================

describe('security limits', () => {
  it('entry_count too large', () => {
    // Build a header with entry_count > MAX_ENTRY_COUNT
    // and total_file_size = HEADER_SIZE so the size check passes
    const buf = Buffer.alloc(64, 0);
    buf.write('SHRD', 0, 'ascii');
    buf[4] = 0x02;           // version
    buf[5] = 1;              // role
    buf.writeUInt16LE(0x00A2, 6); // flags
    buf[8] = 64;             // alignment
    buf[9] = 0;              // compression
    buf.writeUInt16LE(48, 10); // index_entry_size
    buf.writeUInt32LE(MAX_ENTRY_COUNT + 1, 12); // entry_count too large
    buf.writeBigUInt64LE(64n, 16); // string_table_offset
    buf.writeBigUInt64LE(64n, 24); // data_section_offset
    buf.writeBigUInt64LE(0n, 32);  // schema_offset
    buf.writeBigUInt64LE(BigInt(buf.length), 40); // total_file_size
    expect(() => new ShardReader(buf)).toThrow(/MAX_ENTRY_COUNT/);
  });

  it('total_file_size mismatch throws', () => {
    // Make a valid shard then corrupt total_file_size
    const buf = makeShardBuffer([['x', Buffer.from('hello')]]);
    const corrupt = Buffer.from(buf);
    corrupt.writeBigUInt64LE(99999n, 40); // wrong total_file_size
    expect(() => new ShardReader(corrupt)).toThrow(/total_file_size/);
  });

  it('data_offset before data section throws', () => {
    const buf = makeShardBuffer([['x', Buffer.from('hello')]]);
    const corrupt = Buffer.from(buf);
    // Entry 0 data_offset is at HEADER_SIZE + 16 = 80
    corrupt.writeBigUInt64LE(0n, 64 + 16); // set to 0 (before header)
    // Keep total_file_size consistent
    corrupt.writeBigUInt64LE(BigInt(corrupt.length), 40);
    expect(() => new ShardReader(corrupt)).toThrow();
  });

  it('data extends past end of file throws', () => {
    const buf = makeShardBuffer([['x', Buffer.from('hello')]]);
    const corrupt = Buffer.from(buf);
    // Set data_offset to huge value
    corrupt.writeBigUInt64LE(0xFFFFFFFFn, 64 + 16);
    corrupt.writeBigUInt64LE(BigInt(corrupt.length), 40);
    expect(() => new ShardReader(corrupt)).toThrow();
  });
});
